Java: test GeneratedVsManualCoverage query on top 500 JDK APIs

jcogs33 opened the pull request:
Draft PR.

atorralba reviewed (commented):
Mostly looks plausible to me.

OOI, why do are the stubs in `query-tests` instead of `stubs`? And why do we need JDK stubs in the first place?

> but it seems that I can't import `utils.modelgenerator.internal.CaptureModels` in that directory

Yes, that's because `utils` is under `src`, and you can import `lib` modules in `src`, but not the other way around.

jcogs33 commented:
> OOI, why do are the stubs in `query-tests` instead of `stubs`? And why do we need JDK stubs in the first place?

In order to test the query on the top-500 JDK APIs, I needed a database containing those top-500 APIs as callables (since the query is database-dependent and counts APIs as callables). So the stubs are not really being used as stubs here, but rather as the code that the query is being tested against. Let me know if you have a better idea regarding how to set this test case up. :smile:
(Note that this is the same structure as the [original test case](https://github.com/github/codeql/blob/21f9527444a3641c5a2a7b1a968892c99e4ab734/java/ql/test/query-tests/Metrics/GeneratedVsManualCoverage/IOUtils.java) for this query, but at a larger scale since the top-500 JDK API test case is testing against many more models).

> Yes, that's because `utils` is under `src`, and you can import `lib` modules in `src`, but not the other way around.

Ah, sorry, my question wasn't clear; I knew the issue was because `utils.modelgenerator.internal.CaptureModels` is under `src`, but I wanted to make sure that where I ended up placing the `.qll` files due to that issue was okay. I think this issue will actually be resolved once I change this query to not use `DataFlowTargetApi` anymore.

atorralba commented:
> So the stubs are not really being used as stubs here, but rather as the code that the query is being tested against.

Ah, I see! Thanks for the clarification. I was a bit misled by the name, but they _are_ stubs, so it makes sense.  The alternative would be adding code that makes all those calls so that they are extracted, which seems equally bothersome, so I think it's good as it is 😄.

> I wanted to make sure that where I ended up placing the `.qll` files due to that issue was okay.

It looks okay to me. The only thing we should keep in mind is that `TopJdkApis.qll` will now be part of our public API, which means that users will be able to use it, any breaking change will need to go through the deprecation process, and so on. But since you're using it both in `src` and `test`, I don't think we have many alternatives.

jcogs33 commented:
> The only thing we should keep in mind is that `TopJdkApis.qll` will now be part of our public API, which means that users will be able to use it, any breaking change will need to go through the deprecation process, and so on.

Would storing `TopJdkApis.qll` in the `../src/Metrics/Summaries` directory instead avoid this situation? It's importable in the `test` directory when stored there.

Or would it make sense to at least put an `INTERNAL use only.` comment in the file similar to the comment in [`ExternalFlow.qll`](https://github.com/github/codeql/blob/60580efb1680ccee82a95511a34948013c067bcd/java/ql/lib/semmle/code/java/dataflow/ExternalFlow.qll#L2)?

atorralba commented:
> It's importable in the test directory when stored there.

Ah, then I think I slightly favor having it in `src`.

jcogs33 commented:
> Ah, then I think I slightly favor having it in src.

Moved in https://github.com/github/codeql/pull/12680/commits/b534f40b26e3b72b363cb59f568106b58bd18cf0.

Also, I renamed the stubs directory to just `TopJdkApis` avoid confusion with the name.
